metals/src/main/scala/scala/meta/internal/metals/codeactions/ConvertCommentCodeAction.scala

```diff
@@ -0,0 +1,173 @@
+        isSingleLineComment(content.split(LineSeparator).toVector),
+      )
+    } yield codeAction).toList
+  }
+
+  private def tokenizeIfNotRangeSelection(range: l.Range, content: String) = {
+    Option
+      .when(range.getStart == range.getEnd)(
+        Trees.defaultTokenizerDialect(content).tokenize
+      )
+      .flatMap(_.toOption)
+  }
+
+  private def isSingleLineComment(contentLines: Vector[String])(
+      t: Token
+  ): Boolean = t match {
+    case tc: Token.Comment =>
+      val currentLine = contentLines(t.pos.startLine)
+      tc.pos.startLine == tc.pos.endLine &&
+      currentLine.slice(tc.pos.startColumn, tc.pos.startColumn + 2) == "//"
+    case _ => false
+  }
+
+  private def createActionIfPossible(
+      path: AbsolutePath,
+      tokens: Tokens,
+      range: l.Range,
+      isSingleLineComment: Token => Boolean,
+  ): Option[l.CodeAction] = {
+    val indexOfLineTokenUnderCursor = tokens.lastIndexWhere(t =>
+      t.pos.startLine == range.getStart.getLine
+        && t.pos.endLine == t.pos.startLine
+        && t.pos.startColumn < range.getStart.getCharacter
+        && isSingleLineComment(t)
+    )
+    if (indexOfLineTokenUnderCursor != -1) {
+      // tokens that are strictly before cursor, i.e. they end before cursor position
+      val tokensBeforeCursor = tokens.take(indexOfLineTokenUnderCursor)
+      // token under the cursor + following tokens
+      val tokensAfterCursor = tokens.drop(indexOfLineTokenUnderCursor)
+      val textEdit = createTextEdit(
+        tokensBeforeCursor,
+        tokensAfterCursor,
+        range,
+      )
+      Some(
+        CodeActionBuilder.build(
+          title = ConvertCommentCodeAction.Title,
+          kind = this.kind,
+          changes = List(path -> List(textEdit)),
+        )
+      )
+    } else {
+      None
+    }
+  }
+
+  private def createTextEdit(
+      tokensBeforeCursor: Tokens,
+      tokensAfterCursor: Tokens,
+      range: l.Range,
+  ) = {
+    val commentBeforeCursor = collectContinuousComments(
+      tokens = tokensBeforeCursor.reverse
+    ).reverse
+
+    val commentAfterCursor = collectContinuousComments(
+      tokens = tokensAfterCursor
+    )
+    val commentStart = commentBeforeCursor.headOption
+      .map(_.pos.toLsp.getStart)
+      .getOrElse(range.getStart)
+    val commentEnd = commentAfterCursor.lastOption
+      .map(_.pos.toLsp.getEnd())
+      .getOrElse(range.getEnd())
+
+    val commentTokens = commentBeforeCursor ++ commentAfterCursor
+    val replaceText =
+      commentTokens
+        .map(_.value.strip())
+        .mkString("/* ", s"${LineSeparator} * ", " */")
+    if (commentBeforeCursor.isEmpty) {
+      // this is safe as there have to be some tokens after cursor if commentBeforeCursor is empty
+      commentStart.setCharacter(tokensAfterCursor.head.pos.startColumn)
+    }
+    val pos = new l.Range(commentStart, commentEnd)
+    new l.TextEdit(pos, replaceText)
+  }
+
+  private def collectContinuousComments(
```

Comment:
The state machine is nice but I feel like `takeWhile`+ `collect` could achieve the same with less code.

Reply:
Right, I missed the fact that they all extend from `Trivia`. Thanks!